lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -228,25 +257,223 @@ void DominatedAccessRemoval::optimize() {
   }
 }
 
-void DominatedAccessRemoval::perform() {
+// Returns a bool: If we should bail on this function
+// we return false - else true
+bool DominatedAccessRemoval::perform() {
   if (func.empty())
-    return;
+    return false;
 
   if (!analyze()) {
     LLVM_DEBUG(llvm::dbgs()
                << "Bailed on function: " << func.getName() << "\n");
-    return;
+    return false;
   }
   optimize();
+  return true;
+}
+
+namespace {
+class LoopDominatingAccessAdder {
+  SmallVector<SILLoop *, 8> bottomUpWorkList;
+  DominanceInfo *domInfo;
+
+public:
+  LoopDominatingAccessAdder(SILLoop *topLevelLoop, DominanceInfo *domInfo)
+      : domInfo(domInfo) {
+    // Collect loops for a recursive bottom-up traversal in the loop tree.
+    bottomUpWorkList.push_back(topLevelLoop);
+    for (unsigned i = 0; i < bottomUpWorkList.size(); ++i) {
+      auto *loop = bottomUpWorkList[i];
+      for (auto *subLoop : *loop) {
+        bottomUpWorkList.push_back(subLoop);
+      }
+    }
+  }
+
+  void perform();
+
+protected:
+  /// \brief Collect a set of instructions that can be dominated
+  void
+  analyzeCurrentLoop(SILLoop *currnetLoop,
+                     SmallVectorImpl<BeginAccessInst *> &dominatableInstrVec);
+
+  /// \brief Optimize the current loop nest.
+  void optimizeLoop(SILLoop *currnetLoop,
+                    SmallVectorImpl<BeginAccessInst *> &dominatableInstrVec);
+};
+} // namespace
+
+void LoopDominatingAccessAdder::analyzeCurrentLoop(
+    SILLoop *currnetLoop,
+    SmallVectorImpl<BeginAccessInst *> &dominatableInstrVec) {
+  auto *preheader = currnetLoop->getLoopPreheader();
+  if (!preheader) {
+    // Can't dominate this loop - bail
+    return;
+  }
+  for (auto *block : currnetLoop->getBlocks()) {
+    for (auto &instr : *block) {
+      auto *beginAccess = dyn_cast<BeginAccessInst>(&instr);
+      if (!beginAccess) {
+        continue;
+      }
+      if (beginAccess->getEnforcement() != SILAccessEnforcement::Dynamic) {
+        continue;
+      }
+      if (!beginAccess->hasNoNestedConflict()) {
+        continue;
+      }
+      auto operand = beginAccess->getOperand();
+      if (auto *defInstr = operand->getDefiningInstruction()) {
```

Comment:
Use `getParentBlock` instead of duplicating all this code. It works for instructions and arguments.